services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AutoProvisioner.java

```diff
@@ -0,0 +1,386 @@
+                                    // ensure that a notification event gets sent (even if we might send duplicate events)
+                                    return sendDelayedAutoProvisioningNotificationIfNeeded(tenantId, deviceId, gatewayId, deviceData, span)
+                                            .map(deviceData);
+                                });
+                    }
+
+                    span.log("device created");
+                    LOG.trace("device [{}] for gateway [{}] successfully created by auto-provisioning [tenant-id: {}]",
+                            deviceId, gatewayId, tenantId);
+                    return sendAutoProvisioningEvent(tenantId, deviceId, gatewayId, span)
+                            .compose(sendEmptyEventOk -> deviceRegistrationInformationService.processAssertRegistration(DeviceKey.from((TenantKey.from(tenantId)), deviceId), span)
+                                    .compose(deviceRegistrationData -> {
+                                        if (!deviceRegistrationData.isOk()) {
+                                            span.log("update of notification flag failed");
+                                            LOG.warn("notification flag of device [{}] for gateway [{}] of tenant [tenant-id: {}] could not be updated",
+                                                    deviceId, gatewayId, tenantId);
+                                            return Future.failedFuture(StatusCodeMapper.from(deviceRegistrationData.getStatus(),
+                                                    String.format("update of notification flag failed (status %d)", deviceRegistrationData.getStatus())));
+                                        }
+
+                                        final JsonObject deviceData = deviceRegistrationData.getPayload()
+                                                .getJsonObject(RegistrationConstants.FIELD_DATA, new JsonObject());
+                                        return setAutoProvisioningNotificationSent(tenantId, deviceId, deviceData, span)
+                                                .map(deviceData);
+                                    }));
+                })
+                .onFailure(thr -> TracingHelper.logError(span, thr))
+                .onComplete(ar -> span.finish());
+    }
+
+    /**
+     * Notify northbound applications of an auto-provisioned device.
+     *
+     * @param tenantId The id of the tenant for which the edge device should be provisioned.
+     * @param deviceId The id of the edge device which should be provisioned.
+     * @param gatewayId The id of the edge device's gateway.
+     * @param deviceData The data of the edge device.
+     * @param span The span to be used for tracing this operation.
+     *
+     * @return A future indicating the outcome of the operation.
+     *
+     * @see AutoProvisionerConfigProperties#DEFAULT_RETRY_EVENT_SENDING_DELAY
```

Comment:
`AutoProvisionerConfigProperties#DEFAULT_RETRY_EVENT_SENDING_DELAY` is inaccessible from here as it is declared as private.

Reply:
I think it's ok to keep it like this. It doesn't lead to a Javadoc generation error and in an [earlier discussion](https://github.com/eclipse/hono/pull/2118#discussion_r481779045) we kept it as well.

Reply:
It won't throw any error. But in the generated doc, it will show as an invalid link and on clicking it takes to the target class and no information regarding that _private_ field will be available.

Reply:
Then this javadoc part wouldn't be useful in the generated docs, but it would still be useful when browsing the code and reading the javadoc in the IDE.